### PR TITLE
Add writes for channels and channelGroups

### PR DESCRIPTION
**Pull request by alianos-**

### CODE REVIEW

File: src/main/java/org/atlasapi/client/GsonQueryClient.java [hunk 1]
```diff
@@ -340,6 +344,70 @@ public void putPerson(String queryString, Person person) {
...
+        }
+    }
+
+    public ChannelGroupResponse postChannelGroup(String queryString, ChannelGroup channelGroup) {
+        try {
+            String json = gson.get().toJson(channelGroup);
+            Payload httpBody = new StringPayload(json);
+            HttpResponse resp = httpClient.post(queryString, httpBody);
+            if (resp.statusCode() >= 400) {
+                throw new BadResponseException("Error PUTting item: HTTP " + resp.statusCode() + " received from Atlas");
```
Comment (alianos-) [inline src/main/java/org/atlasapi/client/GsonQueryClient.java:369]: POSTing

File: src/main/java/org/atlasapi/client/GsonQueryClient.java [hunk 2]
```diff
@@ -340,6 +344,70 @@ public void putPerson(String queryString, Person person) {
...
+        }
+    }
+
+    public ChannelResponse postChannel(String queryString, Channel channel) {
+        try {
+            String json = gson.get().toJson(channel);
+            Payload httpBody = new StringPayload(json);
+            HttpResponse resp = httpClient.post(queryString, httpBody);
+            if (resp.statusCode() >= 400) {
+                throw new BadResponseException("Error PUTting item: HTTP " + resp.statusCode() + " received from Atlas");
```
Comment (alianos-) [inline src/main/java/org/atlasapi/client/GsonQueryClient.java:401]: deja vu

File: src/main/java/org/atlasapi/client/GsonQueryClient.java [hunk 3]
```diff
@@ -340,6 +344,70 @@ public void putPerson(String queryString, Person person) {
         }
     }
 
+    public ChannelGroupResponse putChannelGroup(String queryString, ChannelGroup channelGroup) {
+        try {
+            String json = gson.get().toJson(channelGroup);
```
Comment (alianos-) [inline src/main/java/org/atlasapi/client/GsonQueryClient.java:349]: Are there tests for this for the other model objects?

Comment (EmiEnache) [reply src/main/java/org/atlasapi/client/GsonQueryClient.java:349]: Nope! Not that I can see anyway

File: src/main/java/org/atlasapi/client/GsonAtlasClient.java [hunk 4]
```diff
@@ -197,6 +201,29 @@ public ContentResponse writePlaylist(Playlist playlist, ContentWriteOptions opti
...
+    }
+
+    @Override
+    public ChannelResponse writeChannel(Channel channel, boolean overwriteExisting) {
+        validateChannel(channel);
+
+        if(overwriteExisting) {
+            return client.putChannel(writeChannelUri(), channel);
+        } else {
+            // TODO postChannel does not exist because there is no POST channel endpoint in owl
```
Comment (tristanbeardmore-gray) [inline src/main/java/org/atlasapi/client/GsonAtlasClient.java:222]: The POST is the one that does exist, the PUT one does not

File: src/main/java/org/atlasapi/client/GsonAtlasClient.java [hunk 5]
```diff
@@ -324,6 +351,18 @@ private String personResourceUri() {
         return queryString;
     }
 
+    private String writeChannelGroupUri() {
+        checkNotNull(apiKey.get(), "An API key must be specified for write queries");
+        String queryString = baseUri + "/channel_groups.json?";
+        return queryString;
```
Comment (tristanbeardmore-gray) [inline src/main/java/org/atlasapi/client/GsonAtlasClient.java:357]: It looks like you need to add the apiKey at this point, if it's anything like the method above

File: src/main/java/org/atlasapi/client/GsonAtlasClient.java [hunk 6]
```diff
@@ -324,6 +351,18 @@ private String personResourceUri() {
...
+    }
+
+    private String writeChannelUri() {
+        checkNotNull(apiKey.get(), "An API key must be specified for write queries");
+        String queryString = baseUri + "/channels.json?";
+        return queryString;
```
Comment (tristanbeardmore-gray) [inline src/main/java/org/atlasapi/client/GsonAtlasClient.java:363]: It looks like you need to add the apiKey at this point, if it's anything like the method above

File: src/main/java/org/atlasapi/client/GsonQueryClient.java [hunk 7]
```diff
@@ -340,6 +344,54 @@ public void putPerson(String queryString, Person person) {
...
+            }
+            WriteResponseWrapper responseWrapper = gson.get().fromJson(resp.body(), WriteResponseWrapper.class);
+            return new ChannelGroupResponse(responseWrapper.getAtlasResponse().getId(), resp.header(LOCATION));
+
+        } catch (HttpException e) {
+            throw Throwables.propagate(e);
+        }
+    }
+
+    public ChannelResponse putChannel(String queryString, Channel channel) {
```
Comment (tristanbeardmore-gray) [inline src/main/java/org/atlasapi/client/GsonQueryClient.java:379]: This needs to be post

File: src/main/java/org/atlasapi/client/GsonQueryClient.java [hunk 8]
```diff
@@ -340,6 +344,54 @@ public void putPerson(String queryString, Person person) {
...
+        try {
+            String json = gson.get().toJson(channel);
+            Payload httpBody = new StringPayload(json);
+            HttpResponse resp = httpClient.put(queryString, httpBody);
```
Comment (tristanbeardmore-gray) [inline src/main/java/org/atlasapi/client/GsonQueryClient.java:383]: This needs to be post

File: src/main/java/org/atlasapi/client/response/ChannelResponse.java [hunk 9]
```diff
@@ -0,0 +1,21 @@
+package org.atlasapi.client.response;
+
+public class ChannelResponse {
```
Comment (tristanbeardmore-gray) [inline src/main/java/org/atlasapi/client/response/ChannelResponse.java:3]: Is this what is actually returned by the endpoint? A quick glance in the code suggests there might not be any response body.

File: src/main/java/org/atlasapi/client/GsonAtlasClient.java (hunk 5 above)
Comment (EmiEnache) [reply src/main/java/org/atlasapi/client/GsonAtlasClient.java:357]: 🤦

File: src/main/java/org/atlasapi/client/response/ChannelGroupResponse.java [hunk 10]
```diff
@@ -0,0 +1,20 @@
+package org.atlasapi.client.response;
+
+public class ChannelGroupResponse {
```
Comment (tristanbeardmore-gray) [inline src/main/java/org/atlasapi/client/response/ChannelGroupResponse.java:3]: Where did you find this type of response? The api seems to produce this response?

`    @JsonCreator
    public WriteResponseWithOldIds(
            @JsonProperty("id") String id,
            @JsonProperty("v4-id") String deerId
    ) {
        this.id = id;
        this.deerId = deerId;
    }`

File: src/main/java/org/atlasapi/client/GsonQueryClient.java [hunk 11]
```diff
@@ -340,6 +344,54 @@ public void putPerson(String queryString, Person person) {
...
+            if (resp.statusCode() >= 400) {
+                throw new BadResponseException("Error PUTting item: HTTP " + resp.statusCode() + " received from Atlas");
+            }
+            WriteResponseWrapper responseWrapper = gson.get().fromJson(resp.body(), WriteResponseWrapper.class);
```
Comment (tristanbeardmore-gray) [inline src/main/java/org/atlasapi/client/GsonQueryClient.java:387]: Are you sure this is the correct response? From a quick glance at the code I'm not even sure there is a response body.

File: src/main/java/org/atlasapi/client/GsonQueryClient.java [hunk 12]
```diff
@@ -340,6 +344,54 @@ public void putPerson(String queryString, Person person) {
...
+
+    public ChannelGroupResponse postChannelGroup(String queryString, ChannelGroup channelGroup) {
+        try {
+            String json = gson.get().toJson(channelGroup);
+            Payload httpBody = new StringPayload(json);
+            HttpResponse resp = httpClient.post(queryString, httpBody);
+            if (resp.statusCode() >= 400) {
+                throw new BadResponseException("Error POSTting item: HTTP " + resp.statusCode() + " received from Atlas");
+            }
+            WriteResponseWrapper responseWrapper = gson.get().fromJson(resp.body(), WriteResponseWrapper.class);
```
Comment (tristanbeardmore-gray) [inline src/main/java/org/atlasapi/client/GsonQueryClient.java:371]: Are you sure this is the right response object? The code seems to construct this:

```    @JsonCreator
    public WriteResponseWithOldIds(
            @JsonProperty("id") String id,
            @JsonProperty("v4-id") String deerId
    ) {
        this.id = id;
        this.deerId = deerId;
    }

File: src/main/java/org/atlasapi/client/GsonQueryClient.java [hunk 13]
```diff
@@ -340,6 +344,54 @@ public void putPerson(String queryString, Person person) {
...
 
+    public ChannelGroupResponse putChannelGroup(String queryString, ChannelGroup channelGroup) {
+        try {
+            String json = gson.get().toJson(channelGroup);
+            Payload httpBody = new StringPayload(json);
+            HttpResponse resp = httpClient.put(queryString, httpBody);
+            if (resp.statusCode() >= 400) {
+                throw new BadResponseException("Error PUTting item: HTTP " + resp.statusCode() + " received from Atlas");
+            }
+            WriteResponseWrapper responseWrapper = gson.get().fromJson(resp.body(), WriteResponseWrapper.class);
```
Comment (tristanbeardmore-gray) [inline src/main/java/org/atlasapi/client/GsonQueryClient.java:355]: Are you sure this is the right response object? The api seems to construct this:

```    @JsonCreator
    public WriteResponseWithOldIds(
            @JsonProperty("id") String id,
            @JsonProperty("v4-id") String deerId
    ) {
        this.id = id;
        this.deerId = deerId;
    }

File: src/main/java/org/atlasapi/client/AtlasWriteClient.java [hunk 14]
```diff
@@ -18,6 +22,10 @@
 
     ContentResponse writePlaylist(Playlist playlist, ContentWriteOptions options);
 
+    ChannelGroupResponse writeChannelGroup(ChannelGroup channelGroup, boolean overwriteExisting);
```
Comment (tristanbeardmore-gray) [inline src/main/java/org/atlasapi/client/AtlasWriteClient.java:25]: For the sake of potential future compatibility I'd use a ChannelGroupWriteOptions object in this signature, like with the ContentWriteOptions. Future options can be added to that builder object without breaking any existing code that way.

File: src/main/java/org/atlasapi/client/AtlasWriteClient.java [hunk 15]
```diff
@@ -18,6 +22,10 @@
...
+
+    ChannelResponse writeChannel(Channel channel, boolean overwriteExisting);
```
Comment (tristanbeardmore-gray) [inline src/main/java/org/atlasapi/client/AtlasWriteClient.java:27]: For the sake of potential future compatibility I'd use a ChannelWriteOptions object in this signature, like with the ContentWriteOptions. Future options can be added to that builder object without breaking any existing code that way.

File: src/main/java/org/atlasapi/client/GsonAtlasClient.java [hunk 16]
```diff
@@ -197,6 +201,29 @@ public ContentResponse writePlaylist(Playlist playlist, ContentWriteOptions opti
...
+    }
+
+    @Override
+    public ChannelResponse writeChannel(Channel channel, boolean overwriteExisting) {
+        validateChannel(channel);
+
+        if(overwriteExisting) {
+            return client.postChannel(writeChannelUri(), channel);
+        } else {
+            // TODO putChannel does not exist because there is no PUT channel endpoint in owl
```
Comment (tristanbeardmore-gray) [inline src/main/java/org/atlasapi/client/GsonAtlasClient.java:222]: This comment now belongs in the other part of the if statement

**Pull request merged**